Lib/dataclasses.py

```diff
@@ -347,20 +347,24 @@ def _create_fn(name, args, body, *, globals=None, locals=None,
     # __builtins__ may be the "builtins" module or
     # the value of its "__dict__",
     # so make sure "__builtins__" is the module.
-    if globals is not None and '__builtins__' not in globals:
-        globals['__builtins__'] = builtins
+    if '__builtins__' not in locals:
+        locals['__builtins__'] = builtins
```

Comment:
This doesn't actually work -- `exec()` gets `__builtins__` only from globals.

Reply:
Probably it's better to rename it to BUILTINS, to make it more obvious.

Reply:
Guido, I've updated the PR, please take a look.